src/ast/ast_validations.c

```diff
@@ -219,308 +129,562 @@ static void _AST_GetProcCallAliases(const cypher_astnode_t *node, rax *identifie
-	uint child_count = cypher_astnode_nchildren(node);
-	for(uint i = 0; i < child_count; i++) {
-		const cypher_astnode_t *child = cypher_astnode_get_child(node, i);
-		cypher_astnode_type_t type = cypher_astnode_type(child);
-
-		if(type == CYPHER_AST_IDENTIFIER) {
-			const char *alias = cypher_ast_identifier_get_name(child);
-			int new = raxInsert(edge_aliases, (unsigned char *)alias, strlen(alias), NULL,
-								NULL);
-			if(!new) {
-				ErrorCtx_SetError("Cannot use the same relationship variable '%s' for multiple patterns.", alias);
-				return AST_INVALID;
-			}
-		}
+// validate an apply-all operator
+static bool _Validate_apply_all_operator
+(
+	const cypher_astnode_t *n,  // ast-node
+	bool start,                 // first traversal
+	ast_visitor *visitor        // visitor
+) {
+	validations_ctx *vctx = visitor->ctx;
+	if(vctx->valid == AST_INVALID || !start) {
+		return false;
 	}
 
-	return AST_VALID;
-}
-
-static AST_Validation _ValidateRelation(rax *projections, const cypher_astnode_t *edge,
-										rax *edge_aliases) {
-	AST_Validation res = AST_VALID;
+	// Working with a function call that has * as its argument.
+	const cypher_astnode_t *func = cypher_ast_apply_all_operator_get_func_name(n);
+	const char *func_name = cypher_ast_function_name_get_value(func);
 
-	// Make sure edge alias is unique
-	res = _Validate_ReusedEdges(edge, edge_aliases);
-	if(res != AST_VALID) return res;
+	// Verify that this is a COUNT call.
+	if(strcasecmp(func_name, "COUNT")) {
+		ErrorCtx_SetError("COUNT is the only function which can accept * as an argument");
+		vctx->valid = AST_INVALID;
+		return false;
+	}
 
-	// If this is a multi-hop traversal, validate it
-	const cypher_astnode_t *range = cypher_ast_rel_pattern_get_varlength(edge);
-	if(range) {
-		res = _ValidateMultiHopTraversal(projections, edge, range);
-		if(res != AST_VALID) return res;
+	// Verify that DISTINCT is not specified.
+	if(cypher_ast_apply_all_operator_get_distinct(n)) {
+		// TODO consider opening a parser error, this construction is invalid in Neo's parser.
+		ErrorCtx_SetError("Cannot specify both DISTINCT and * in COUNT(DISTINCT *)");
+		vctx->valid = AST_INVALID;
+		return false;
 	}
 
-	return res;
+	return true;
 }
 
-static AST_Validation _ValidatePath
+// validate an apply operator
+static bool _Validate_apply_operator
 (
-	const cypher_astnode_t *path,
-	rax *projections,
-	rax *edge_aliases
+	const cypher_astnode_t *n,  // ast-node
+	bool start,                 // first traversal
+	ast_visitor *visitor        // visitor
 ) {
-	AST_Validation res = AST_VALID;
+	validations_ctx *vctx = visitor->ctx;
+	if(vctx->valid == AST_INVALID || !start) {
+		return false;
+	}
+
+	// Collect the function name.
+	const cypher_astnode_t *func = cypher_ast_apply_operator_get_func_name(n);
+	const char *func_name = cypher_ast_function_name_get_value(func);
+	vctx->valid = _ValidateFunctionCall(func_name, vctx->clause == CYPHER_AST_WITH 
```

Comment:
`return _ValidateFunctionCall(func_name, vctx->clause == CYPHER_AST_WITH || vctx->clause == CYPHER_AST_RETURN);`

the logic responsible for the traversal should conclude with a failure once a visit function returns false.